### ISSUE TITLE
setLedTxt not parsed correctly, existing macros from UHK60 throw errors

### ISSUE THREAD

**Issue by mhantsch** [UltimateHackingKeyboard/firmware#933]
My macros that I imported from UHK60 contain many `setLedTxt` commands. All of them throw errors at the first parameter (the time), not the actual setLedTxt command.

```
Error at $onInit 1/24/11: Unrecognized command: 300
> 24 | [setLedTxt 300 ':::'](https://github.com/UltimateHackingKeyboard/agent80/issues/new#/macro/0?actionIndex=0&lineNr=24&columnNr=11&inlineEdit=true)
>    |           ^
Error at $onKeymapChange CMX 1/2/11: Unrecognized command: 1
> 2 | [setLedTxt 1 CMX](https://github.com/UltimateHackingKeyboard/agent80/issues/new#/macro/2?actionIndex=0&lineNr=2&columnNr=11&inlineEdit=true)
>   |           ^
```

**Comment by kareltucek** [comment]
What firmware commit do you have flashed?

My own config contains setLedTxt many times, and I am getting no errors.

**Comment by mhantsch** [comment]
Hadn't flashed anything new yet, it's on whatever Laszlo had on it when I picked it up this Monday. Agent says `UltimateHackingKeyboard/firmware-uhk80: 11.1.0 #1ade306f`.

**Comment by mhantsch** [comment]
Tried to update to whatever this Agent80 came with, but doesn't work:

![image](https://github.com/user-attachments/assets/e22eb84a-4e49-4aad-8078-6889d3098250)

**Comment by mondalaci** [comment]
Agent 80 bundles the stock firmware. We don't fix this because the Agent 80 and firmware-uhk80 repos will soon be merged into the stock repos. In the meantime, please grab firmware tarballs at https://github.com/UltimateHackingKeyboard/firmware-uhk80/actions

**Comment by kareltucek** [comment]
That's strange. That's a pretty recent commit.

Anyways, at this moment this a a low priority issue.

**Comment by mhantsch** [comment]
Will retry with the latest commit tarball in the next few days.

**Comment by mhantsch** [comment]
I tried after downloading uhk-firmware-11.1.0-42d6d1d.tar.gz but this is what I get:

```
Start flashing firmware
[DeviceService] Operating system: Linux 6.9.3-76060903-generic x64
[DeviceService] Agent version: 4.2.0
[DeviceService] New firmware version: 11.1.0
[DeviceService] New firmware user config version: 7.1.0
[DeviceService] stop poll UHK Device
[DeviceService] stopped poll UHK Device
[DeviceOperation] right module deviceProtocolVersion: 4.10.0
[DeviceOperation] Read "leftHalf" version information
[DeviceOperation] Read "leftHalf" repo information
[DeviceOperation] Read "leftHalf" firmware checksum
[DeviceService] UHK Device firmware upgrade starts: {"id":4,"firmwareUpgradeMethod":"mcuboot","logName":"UHK 80 right","name":"UHK 80","keyboard":[{"vid":14248,"pid":9}],"bootloader":[{"vid":14248,"pid":8}],"buspal":[],"reportId":4}
[DeviceService] Device right firmware version: 11.1.0
[UhkOperations] Start flashing device firmware with mcuboot
[UhkOperations] Start flashing UHK 80 right firmware with mcumgr
[UhkOperations] Reenumerate bootloader
[UhkHidDevice] Device communication closing.
[UhkHidDevice] Device communication closed.
[UhkHidDevice] Start reenumeration, mode: Bootloader, timeout: 5000ms
[UhkHidDevice] try to find MCU Bootloader
[UhkHidDevice] Reenumeration error. We hope it would not break the process TypeError: Cannot write to hid device
    at Object.reenumerate (/tmp/.mount_UHK.AghQFM4d/resources/app.asar/electron-main.js:2:612617)
    at async Object.updateFirmwareWithMcuManager (/tmp/.mount_UHK.AghQFM4d/resources/app.asar/electron-main.js:2:619057)
    at async es.updateFirmware (/tmp/.mount_UHK.AghQFM4d/resources/app.asar/electron-main.js:2:509195)
    at async Tn.process (/tmp/.mount_UHK.AghQFM4d/resources/app.asar/electron-main.js:2:481758)
[UhkHidDevice] try to find MCU Bootloader
[UhkHidDevice] Could not find reenumerated device: Bootloader. Waiting...
[UhkHidDevice] Available devices changed.
[UhkHidDevice] Removed: {"vendorId":"0x37A8","productId":"0x9","path":"/dev/hidraw3","serialNumber":"1736713552","manufacturer":"Ultimate Gadget Laboratories","product":"UHK 80 right half","release":1,"interface":0,"usagePage":1,"usage":6}
[UhkHidDevice] Removed: {"vendorId":"0x37A8","productId":"0x9","path":"/dev/hidraw4","serialNumber":"1736713552","manufacturer":"Ultimate Gadget Laboratories","product":"UHK 80 right half","release":1,"interface":1,"usagePage":1,"usage":2}
[UhkHidDevice] Removed: {"vendorId":"0x37A8","productId":"0x9","path":"/dev/hidraw6","serialNumber":"1736713552","manufacturer":"Ultimate Gadget Laboratories","product":"UHK 80 right half","release":1,"interface":2,"usagePage":65280,"usage":1}
[UhkHidDevice] Removed: {"vendorId":"0x37A8","productId":"0x9","path":"/dev/hidraw7","serialNumber":"1736713552","manufacturer":"Ultimate Gadget Laboratories","product":"UHK 80 right half","release":1,"interface":3,"usagePage":12,"usage":1}
[UhkHidDevice] Removed: {"vendorId":"0x37A8","productId":"0x9","path":"/dev/hidraw8","serialNumber":"1736713552","manufacturer":"Ultimate Gadget Laboratories","product":"UHK 80 right half","release":1,"interface":4,"usagePage":1,"usage":5}
[UhkHidDevice] try to find MCU Bootloader
[UhkHidDevice] Could not find reenumerated device: Bootloader. Waiting...
[UhkHidDevice] try to find MCU Bootloader
[UhkHidDevice] Could not find reenumerated device: Bootloader. Waiting...
[UhkHidDevice] try to find MCU Bootloader
[UhkHidDevice] Could not find reenumerated device: Bootloader. Waiting...
[UhkHidDevice] try to find MCU Bootloader
[UhkHidDevice] Could not find reenumerated device: Bootloader. Waiting...
[UhkHidDevice] try to find MCU Bootloader
[UhkHidDevice] Could not find reenumerated device: Bootloader. Waiting...
[UhkHidDevice] try to find MCU Bootloader
[UhkHidDevice] Could not find reenumerated device: Bootloader. Waiting...
[UhkHidDevice] try to find MCU Bootloader
[UhkHidDevice] Could not find reenumerated device: Bootloader. Waiting...
[UhkHidDevice] try to find MCU Bootloader
[UhkHidDevice] Device communication closing.
[UhkOperations] Init SerialPeripheral: /dev/ttyACM0
[UhkOperations] Read UHK 80 right firmware from file
[UhkOperations] Write memory with mcumgr
[DeviceService] updateFirmware error {"message":"Error: Permission denied, cannot open /dev/ttyACM0","stack":"Error: Error: Permission denied, cannot open /dev/ttyACM0"}
[UhkHidDevice] Available devices unchanged
[DeviceService] Read hardware modules information failed Error: [UhkHidDevice] Device is not connected
    at /tmp/.mount_UHK.AghQFM4d/resources/app.asar/electron-main.js:2:610242
    at new Promise (<anonymous>)
    at Object.write (/tmp/.mount_UHK.AghQFM4d/resources/app.asar/electron-main.js:2:610176)
    at Object.getDeviceState (/tmp/.mount_UHK.AghQFM4d/resources/app.asar/electron-main.js:2:613802)
    at Object.waitUntilKeyboardBusy (/tmp/.mount_UHK.AghQFM4d/resources/app.asar/electron-main.js:2:625026)
    at es.getHardwareModules (/tmp/.mount_UHK.AghQFM4d/resources/app.asar/electron-main.js:2:506115)
    at es.updateFirmware (/tmp/.mount_UHK.AghQFM4d/resources/app.asar/electron-main.js:2:512433)
    at async Tn.process (/tmp/.mount_UHK.AghQFM4d/resources/app.asar/electron-main.js:2:481758)
[DeviceService] start poll UHK Device
Error: Permission denied, cannot open /dev/ttyACM0
```

And after this, only the right half returns. Unplugging and replugging got both halves back again, but after the next attempt I had to restart them.

Anyway, it looks like I cannot update the firmware right now.

**Comment by mondalaci** [comment]
Agent should have updated your udev rules. Please paste the output of `cat /etc/udev/rules.d/50-uhk60.rules`.

**Comment by mhantsch** [comment]
```
# Ultimate Hacking Keyboard rules
# These are the udev rules for accessing the USB interfaces of the UHK as non-root users.
# Copy this file to /etc/udev/rules.d and physically reconnect the UHK afterwards.
SUBSYSTEM=="input", ATTRS{idVendor}=="1d50", ATTRS{idProduct}=="612[0-7]", GROUP="input", MODE="0660"
SUBSYSTEMS=="usb", ATTRS{idVendor}=="1d50", ATTRS{idProduct}=="612[0-7]", TAG+="uaccess"
KERNEL=="hidraw*", ATTRS{idVendor}=="1d50", ATTRS{idProduct}=="612[0-7]", TAG+="uaccess"
```

**Comment by mhantsch** [comment]
Also, my HRMs aren't working correctly on the UHK80. I think the `ifShortcut` command seems to behave differently, but I can't really debug it because `setLedTxt` isn't working. In any case, I think `setLedTxt` must be made compatible and produce it's output on the OLED display.

**Comment by mondalaci** [comment]
This is the old udev rule file which should have been updated. Are you running the latest Agent 80? It should update your udev file.

**Comment by mhantsch** [comment]
I am running this one:

(base) max@max-framework:~$ sha512sum ~/Downloads/UHK.Agent-4.2.0-linux-x86_64.AppImage 
35bea18747ba7bb338df64e1a9e53eb7e0e71b7847f7cbda356b2b67bf04c340b0a78a7165d807c2f45f438ba48d9f58dc8dc554fa9a2ff09d2ddded7d6c5da5  /home/max/Downloads/UHK.Agent-4.2.0-linux-x86_64.AppImage

**Comment by kareltucek** [comment]
Yeah, I know, and will appreciate if you stop complaining about minor defects of the macro engine until we manage to implement basic hardware support 😉.

**Comment by mhantsch** [comment]
I can confirm the setLedTxt parsing issue is now fixed (tested with firmware-uhk80 11.1.0 #64396d1).
